state/src/caching/order_metadata_index.rs

```diff
@@ -0,0 +1,263 @@
+//! Indexing of orders by metadata
+//!
+//! Specifically, we build an efficiently queryable mapping from pair -> side ->
+//! (matchable_amount, order_id)
+//!
+//! This allows the matching engine to efficiently query a narrow set of
+//! candidate orders to match against a target order.
+
+use std::collections::HashMap;
+
+use circuit_types::{order::OrderSide, Amount};
+use common::types::wallet::{Order, OrderIdentifier, Pair};
+use tokio::sync::RwLock;
+
+use super::RwLockHashMap;
+
+/// A type alias for a the inner index mapping
+type OrderSideIndex = HashMap<OrderSide, SortedVec<(Amount, OrderIdentifier)>>;
+
+/// The order metadata index
+#[derive(Default)]
+pub struct OrderMetadataIndex {
+    /// The mapping from pair -> side -> (matchable_amount, order_id)
+    index: RwLockHashMap<Pair, OrderSideIndex>,
+}
+
+impl OrderMetadataIndex {
+    /// Construct a new order metadata index
+    pub fn new() -> Self {
+        let index = RwLock::new(HashMap::new());
+        Self { index }
+    }
+
+    // --- Getters --- //
+
+    /// Get all orders for a given pair and side, sorted by matchable amount
+    pub async fn get_orders(&self, pair: &Pair, side: OrderSide) -> Vec<OrderIdentifier> {
+        let index = self.index.read().await;
+        let orders_vec = index.get(pair).and_then(|side_index| side_index.get(&side));
+        match orders_vec {
+            Some(v) => v.iter().map(|(_, oid)| *oid).collect(),
+            None => Vec::new(),
+        }
+    }
+
+    // --- Setters --- //
+
+    /// Add an order to the index
+    pub async fn add_order(
+        &self,
+        order_id: OrderIdentifier,
+        order: &Order,
+        matchable_amount: Amount,
```

Comment:
what's the plan for how we compute this? for buy orders, this value depends on the price. will we fetch new prices and update the index on some interval?

Reply:
it won't be converted via price. so if the order is a buy order, we will just make this the amount of the quote token held by the wallet.